### PR TITLE
Association

### PR DESCRIPTION
**Pull request by ANTOINE1128**
# in this PR I have followed these project requirements :
 ## Create a class Classroom with the following:
@label instance variable should be initialized in the constructor.
Setter and getter for @label (remember about attr_accessor).
 ## Create the has-many/belongs-to relationship between Classroom and Student. The following should be implemented:
 ## Create the has-many side (a classroom has many students).
 ## Create the belongs-to side (a student belongs to a classroom).
Make sure that when adding a student to a classroom it also sets the classroom for the student.
Make sure that when setting the classroom for a student it also adds it to the classrooms' students.
 ## Create a class Book with the following:
@title and @author instance variables should be initialized in the constructor.
Setters and getters for instance variables (remember about attr_accessor). 
 ## Create a class Rental with the following:
@date instance variable should be initialized in the constructor.
Setter and getter for @date (remember about attr_accessor).
 ## Create the many-to-many (also has-many-through) relationship between Person and Book using the intermediate class Rental. The following should be implemented:
 ## Create the has-many side of Book and Rental (a book has many rentals).
 ## Create the belongs-to side of Rental and Book (a rental belongs to a book).
 ## Create the has-many side of Person and Rental (a person has many rentals).
 ## Create the belongs-to side of Rental and Person (a rental belongs to a person).
Modify the constructor of Rental so Book and Person are set in it.

### CODE REVIEW

File: book.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  attr_accessor :title, :author, :date
+  attr_reader :rentals
+
+  def initialize(title = 'no title', author = 'no author')
+    @title = title
+    @author = author
+    @rentals = []
+  end
+
+  def add_rental(rental)
```
Comment (souad988) [inline book.rb]: Good job by adding add_rental method to the book class 👍🏼  , but the optimal solution to handle this many to many relation is to create the new rental instance here in this method and in rental initializer push in rentals of book and of person this way when you call Book.add_rental all 'belongs-to' and has-many associations will be created

```ruby
def add_rental(person, date)
    # *the book itself should be sent as a parameter to create the new rental achieving the 'has-many' association. 
    Rental.new(date, self, person)
  end
```

File: person.rb [hunk 2]
```diff
@@ -1,24 +1,31 @@
...
     @name = name
     @age = age
     @parent_permission = parent_permission
-    super()
+    @rentals = []
+  end
+
+  def add_rental(rental)
+    @rentals << rental
   end
```
Comment (souad988) [inline person.rb]: same comment as in Book add_rental goes here

File: person.rb [hunk 3]
```diff
@@ -1,24 +1,31 @@
 require_relative 'nameable'
-require_relative 'Decorator'
-require_relative 'Trimmer_Decorator'
-require_relative 'Capitalize_Decorator'
+# require_relative 'Decorator'
+# require_relative 'Trimmer_Decorator'
+# require_relative 'Capitalize_Decorator'
+# require_relative 'classroom'
+# require_relative 'student'
```
Comment (souad988) [inline person.rb]: kindly make sure to remove all commented code to keep your code base clean

File: student.rb [hunk 4]
```diff
@@ -1,10 +1,13 @@
...
 class Student < Person
-  def initialize(classroom: 'no classroom', name: 'Unknown', age: nil, parent_permission: true)
-    super(name: name, age: age, parent_permission: parent_permission)
+  attr_accessor :classroom, :name
+
+  def initialize(classroom, age, name = 'Unknown', parent_permission: true)
+    super(age, name, parent_permission: parent_permission)
+    # @name = name
     @classroom = classroom
+    classroom.students.push(self) unless classroom.students.include?(self)
```
Comment (souad988) [inline student.rb:10]: I would suggest creating a setter for classroom attribute as we may need to create a student before assigning a classroom 
Check for a method that serves as the 'belongs-to' side (a student belongs to a classroom). [https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#both-ways-relationships](https://github.com/microverseinc/curriculum-ruby/blob/main/oop/articles/oop_relationships_by_examples.md#both-ways-relationships)

````ruby
attr_reader :classroom
  def classroom=(classroom)
      @classroom = classroom
      classroom.students.push(self) unless classroom.students.include?(self)
  end
````

**Review by souad988 — CHANGES_REQUESTED** [review]
Hi @ANTOINE1128 , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


242722065-db428b0c-8dfc-49c4-a156-42f227e32868


## Highlights
Professional README
Add exemples to run code

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by tufoinnkuo10 — APPROVED** [review]
## Status: Approved :heavy_check_mark:

Hi @ANTOINE1128 

Your project is complete! Great job for implementing the requested changes🥇. There is nothing else to say other than... it's time to merge it 🍾🚢 :shipit: . Congratulations!  🎉💯🌟 

## Good Points 👍
- Descriptive Pull Request
- classes well linked
- Methods ok
- No linter errors

<br>
Cheers..Happy coding!..and keep soaring higher! 💻 🍷 🚀 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have 2 more limited reviews per this project. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**